File: weeks/6/4/solution.cpp
#include <iostream>

using namespace std;

constexpr int SIDE = 1001;

int grid[SIDE][SIDE];

void solve() {
  int numSpecies, threshold; // N, K
  cin >> numSpecies >> threshold;

  // 각 종마다 영역의 꼭지점에 서식 영역 인코딩
  for (int i = 0; i < numSpecies; ++i) {
    int x1, y1, x2, y2;
    cin >> x1 >> y1 >> x2 >> y2;
    grid[y1][x1] += 1;
    grid[y1][x2] -= 1;
    grid[y2][x1] -= 1;
    grid[y2][x2] += 1;
  }

  // 행에 대한 프리픽스 섬
  for (int y = 0; y < SIDE; ++y) {
    int acc = 0;
    for (int x = 0; x < SIDE; ++x) {
      acc += grid[y][x];
      grid[y][x] = acc;
    }
  }

  // 열에 대한 프리픽스 섬
  for (int x = 0; x < SIDE; ++x) {
    int acc = 0;
    for (int y = 0; y < SIDE; ++y) {
      acc += grid[y][x];
      grid[y][x] = acc;
    }
  }

  // 종이 threshold 만큼 있는 칸의 개수 세기
  int count = 0;
  for (int y = 0; y < SIDE; ++y) {
    for (int x = 0; x < SIDE; ++x) {
      if (grid[y][x] == threshold) {
        count++;
      }
    }
  }

  cout << count << '\n';
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  solve();
}

File: weeks/6/4/solution.py
import sys
sys.setrecursionlimit(100000)
readline = sys.stdin.readline

SIDE = int(1e3+1)

grid = [[0] * SIDE for _ in range(SIDE)]

numSpecies, threshold = map(int, readline().split())

# 각 종마다 영역의 꼭지점에 서식 영역 인코딩
for _ in range(numSpecies):
    x1, y1, x2, y2 = map(int, readline().split())
    grid[y1][x1] += 1
    grid[y1][x2] -= 1
    grid[y2][x1] -= 1
    grid[y2][x2] += 1

# 행에 대한 프리픽스 섬
for y in range(0, SIDE):
    for x in range(1, SIDE):
        grid[y][x] += grid[y][x-1]

# 열에 대한 프리픽스 섬
for x in range(0, SIDE):
    for y in range(1, SIDE):
        grid[y][x] += grid[y-1][x]

# 종이 threshold 만큼 있는 칸의 개수 세기
count = 0
for y in range(SIDE):
    for x in range(SIDE):
        if grid[y][x] == threshold:
            count += 1

print(count)
